// truffle-decoder/dist/decode/mapping.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = __importDefault(require("./index"));
function decodeMapping(definition, pointer, info, web3, contractAddress) {
    if (definition.referencedDeclaration) {
        // attempting to decode reference to mapping, thus missing valid pointer
        return undefined;
    }
    const { mappingKeys } = info;
    // debug("mapping %O", pointer);
    // debug("mapping definition %O", definition);
    let keys = mappingKeys[definition.id] || [];
    // debug("known keys %o", keys);
    let keyDefinition = definition.typeName.keyType;
    let valueDefinition = definition.typeName.valueType;
    let baseSlot = pointer.storage.from.slot;
    let mapping = {};
    // debug("mapping %O", mapping);
    for (let key of keys) {
        let keyPointer = {
            literal: key
        };
        let valuePointer = {
            storage: {
                from: {
                    slot: {
                        key: key,
                        path: baseSlot.path || undefined,
                        offset: baseSlot.offset
                    },
                    index: 0
                },
                to: {
                    slot: {
                        key: key,
                        path: baseSlot.path || undefined,
                        offset: baseSlot.offset
                    },
                    index: 31
                }
            }
        };
        // debug("keyPointer %o", keyPointer);
        // NOTE mapping keys are potentially lossy because JS only likes strings
        let keyValue = index_1.default(keyDefinition, keyPointer, info, web3, contractAddress);
        // debug("keyValue %o", keyValue);
        if (keyValue != undefined) {
            mapping[keyValue.toString()] =
                index_1.default(valueDefinition, valuePointer, info, web3, contractAddress);
        }
    }
    return mapping;
}
exports.default = decodeMapping;
